### PR TITLE
Fill in VersioningService and usages

### PR DESCRIPTION
**Pull request by NamelessCoder**
Adds needed support for:
- Check-in
- Check-out
- Cancel checked-out
- Get all versions
- Get latest version
- Get latest version's properties

Changes VersioningServiceInterface to make checkIn() and checkOut() return values.

### CODE REVIEW

**Comment by NamelessCoder** [comment]
Question open for discussion:

The changes performed in this commit change the current expectations for the VersioningServiceInterface methods checkIn() and checkOut(), making them return new object IDs. To reflect this a simple change in method annotation is sufficient.

The reason this is done is to avoid changing the PHP signature of those two methods (and the implementation) to pass object ID by reference (as is the strategy in other places). Such a change is potentially breaking towards implementations that are not our own - but given the young age of the project it may be preferable to change the signatures _now_ for consistency.

Arguments for and against are welcome!

**Comment by sascha-egerer** [comment]
HI @NamelessCoder, i can remember that we already talked about that reference stuff. The Problem is that in the java implmementation a "holder" object exists which is just there to pass the argument as reference. I really don't like passing by reference but i think it's ok at that place. I would try to keep the code as consistend as possible and as near as possible to the java implementation (https://github.com/apache/chemistry-opencmis/blob/ac572d9a4d8d13bf5caed0179fc1b89be0ec9bbf/chemistry-opencmis-client/chemistry-opencmis-client-bindings/src/main/java/org/apache/chemistry/opencmis/client/bindings/spi/local/VersioningServiceImpl.java)
So i would vote for passing by reference even if that is a breaking change now... there is no official version released yet so i don't see a reason why we should not do breaking changes...

**Comment by NamelessCoder** [comment]
Don't merge yet - change is coming to implement the alternative solution described above, changing the signature to pass object ID by reference and changing this referenced value inside the function.

**Comment by NamelessCoder** [comment]
Just a rebase for now

File: src/DataObjects/Document.php [hunk 1]
```diff
@@ -375,10 +375,10 @@ public function getAllVersions(OperationContextInterface $context = null)
 
         $objectFactory = $this->getSession()->getObjectFactory();
         $result = array();
-        if (count($versions) !== null) {
+        if (count($versions)) {
             foreach ($versions as $objectData) {
                 $document = $objectFactory->convertObject($objectData, $context);
-                if (!(!$document instanceof DocumentInterface)) {
+                if (!($document instanceof DocumentInterface)) {
                     // this should never happen
```
Comment (ksjogo) [inline src/DataObjects/Document.php]: Could we use map to make these lines clearer in purpose?
When won't document instanceof DocumentInterface fulfill?

Comment (NamelessCoder) [inline src/DataObjects/Document.php]: `convertObject` may return a non-Document and this clause guards against that. However, if such a response is encountered it is most definitely a (big) problem with the repository implementation - it is very much unexpected that the repository would return anything other than Documents when requesting versions of a Document which is the only resource that supports versions. Replaced comment with throwing of exception.

**Comment by NamelessCoder** [comment]
PR now changed to implement the second proposed solution with changing the VersioningService interface and implementation to pass object ID by reference. Should be ready to merge.

**Comment by NamelessCoder** [comment]
@sascha-egerer If you've got a couple of minutes would you mind giving this one a second read? Updated to pass by reference as we discussed ;)

File: src/Bindings/Browser/VersioningService.php [hunk 2]
```diff
@@ -31,9 +32,18 @@ class VersioningService extends AbstractBrowserBindingService implements Version
      * @param string $objectId the identifier for the PWC
      * @param ExtensionDataInterface|null $extension
      */
-    public function cancelCheckOut($repositoryId, $objectId, ExtensionDataInterface $extension = null)
+    public function cancelCheckOut($repositoryId, & $objectId, ExtensionDataInterface $extension = null)
```
Comment (sascha-egerer) [inline src/Bindings/Browser/VersioningService.php:35]: `& $objectId` must be `&$objectId`

File: src/Bindings/Browser/VersioningService.php [hunk 3]
```diff
@@ -56,7 +66,7 @@ public function cancelCheckOut($repositoryId, $objectId, ExtensionDataInterface
      */
     public function checkIn(
         $repositoryId,
-        $objectId,
+        & $objectId,
```
Comment (sascha-egerer) [inline src/Bindings/Browser/VersioningService.php:69]: same as above

Comment (NamelessCoder) [inline src/Bindings/Browser/VersioningService.php:69]: Style is copied from https://github.com/dkd/php-cmis-client/blob/master/src/Bindings/Browser/VersioningService.php#L84 (should we change that across the repository in another commit?)

**Comment by sascha-egerer** [comment]
Looks good by reading. Just would change some formatting but i've notices that we also used the format `& $` on other places... So... We can also change that in a different patch

**Comment by NamelessCoder** [comment]
Roger that - I'll take care of that `&` change next in a new PR!

**Comment by ksjogo** [comment]
```
-*- mode: compilation; default-directory: "~/fi/env/vendor/dkd/php-cmis/examples/" -*-
Compilation started at Wed May  4 18:24:11

php ./Versioning.php 1
PHP Warning:  Module 'pcntl' already loaded in Unknown on line 0

Warning: Module 'pcntl' already loaded in Unknown on line 0
[*] Using existing README.md document in repository root folder.
[*] Checking out document to private working copy (PWC)... done!
[*] Versioned ID before: 6fdf8d88-3597-420b-a2b0-aa67a2970949;3.0
[*] Versioned ID during checkout: 6fdf8d88-3597-420b-a2b0-aa67a2970949;pwc
[*] Checking in with updated properties (major)
PHP Fatal error:  Call to a member function getId() on null in /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/src/DataObjects/Document.php on line 139
PHP Stack trace:
PHP   1. {main}() /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/examples/Versioning.php:0
PHP   2. Dkd\PhpCmis\DataObjects\Document->checkIn() /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/examples/Versioning.php:100

Fatal error: Call to a member function getId() on null in /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/src/DataObjects/Document.php on line 139

Call Stack:
    0.0016     249440   1. {main}() /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/examples/Versioning.php:0
    1.1921    8317832   2. Dkd\PhpCmis\DataObjects\Document->checkIn() /Users/dkd-goslar/fi/env/vendor/dkd/php-cmis/examples/Versioning.php:100


Compilation exited abnormally with code 255 at Wed May  4 18:24:13
```

**Comment by ksjogo** [comment]
Caused by the reference/return mixture?

**Pull request closed without merging**